src/openrct2/drawing/drawing_fast.cpp

```diff
@@ -36,104 +31,119 @@ static void FASTCALL DrawRLESprite2(const uint8* RESTRICT source_bits_pointer,
+            if (x_start + numPixels > width)
+                numPixels = width - x_start;
+
+            uint8 *copyDest = loop_dest_pointer + (x_start >> zoom_level);
 
             //Finally after all those checks, copy the image onto the drawing surface
             //If the image type is not a basic one we require to mix the pixels
-            if (image_type & IMAGE_TYPE_REMAP) {//In the .exe these are all unraveled loops
-                for (; no_pixels > 0; no_pixels -= zoom_amount, source_pointer += zoom_amount, dest_pointer++) {
-                    uint8 al = *source_pointer;
-                    uint8 ah = *dest_pointer;
+            if (image_type & IMAGE_TYPE_REMAP)  // palette controlled images
+            {
+                for (int j = 0; j < numPixels; j += zoom_amount, copySrc += zoom_amount, copyDest++)
+                {
                     if (image_type & IMAGE_TYPE_TRANSPARENT)
-                        al = palette_pointer[(((uint16)al << 8) | ah) - 0x100];
+                    {
+                        uint8 color = (((uint16)*copySrc << 8) | *copyDest) - 0x100;
+                        *copyDest = palette_pointer[color];
+                    }
                     else
-                        al = palette_pointer[al];
-                    *dest_pointer = al;
+                    {
+                        *copyDest = palette_pointer[*copySrc];
+                    }
                 }
-            } else if (image_type & IMAGE_TYPE_TRANSPARENT) {//In the .exe these are all unraveled loops
-                //Doesn't use source pointer ??? mix with background only?
-                //Not Tested
-
-                for (; no_pixels > 0; no_pixels -= zoom_amount, dest_pointer++) {
-                    uint8 pixel = *dest_pointer;
+            }
+            else if (image_type & IMAGE_TYPE_TRANSPARENT)  // single alpha blended color (used for glass)
+            {
+                for (int j = 0; j < numPixels; j += zoom_amount, copyDest++)
+                {
+                    uint8 pixel = *copyDest;
                     pixel = palette_pointer[pixel];
-                    *dest_pointer = pixel;
+                    *copyDest = pixel;
                 }
-            } else
+            }
+            else  // standard opaque image
             {
-                if (zoom_amount == 1) {
-                    no_pixels &= ~less_or_equal_zero_mask(no_pixels);
-                    memcpy(dest_pointer, source_pointer, no_pixels);
-                } else {
-                    for (; no_pixels > 0; no_pixels -= zoom_amount, source_pointer += zoom_amount, dest_pointer++) {
-                        *dest_pointer = *source_pointer;
-                    }
+                if (zoom_level == 0)
+                {
+                    // Since we're sampling each pixel at this zoom level, just do a straight memcpy
+                    if (numPixels > 0)
+                        memcpy(copyDest, copySrc, numPixels);
```

Comment:
`memcpy` allows 0, and probably already includes the check - so you can remove the if statement.